### PR TITLE
bounds checking.

### PR DESCRIPTION
**Pull request by bebbi**
Every expression checks its max bounds. Non-leaf nodes evaluate downstream expressions. Expressions with no foreseeable bounds have a bound of infinity.
on first check, expression caches bounds for performance.

### CODE REVIEW

**Comment by bebbi** [comment]
Is this along your #3 idea? I think it could benefit from some review. See also tests.

**Comment by bebbi** [comment]
ps just realized that we can cache bounds when I wanted to pull. With this, the rotation bounds can probably be refined - maybe you have a better approach for that.

File: lib/expressions.js [hunk 1]
```diff
@@ -140,6 +206,28 @@ expressions.rotate = {
...
     this.matrix.makeRotationAxis(v, this.angle);
   },
+  bounds: function() {
+    // find max distance to rotation axis, expand cube to that *sqrt(2) around axis.
+    // this could likely be optimized, but calc trade-off?
+    if (this.angle === 0) {
+      // trivial
+      return this.operands[0].bounds();
+    }
+    // TODO: introduce rotation center rotation? what if this.origin is not 0?
```
Comment (pieper) [inline lib/expressions.js:216]: Let's leave the center of rotation at the origin.  Other centers can be a translation followed by rotation followed by translation back.

File: lib/expressions.js [hunk 2]
```diff
@@ -140,6 +206,28 @@ expressions.rotate = {
...
+    var origin = [0, 0, 0];
+    var distances = [];
+    for (var b = 0; b < 2; b++) {
+      for (var i = 0; i < 3; i++) {
+        distances.push(Math.abs(this.operands[0].bounds()[b][i] - origin[i]));
+      }
+    }
+    // factor sqrt(2) for Pi/4 turn of a cube
+    var maxd = Math.max.apply(this, distances) * Math.sqrt(2);
```
Comment (pieper) [inline lib/expressions.js:225]: Would be better to apply the inverse rotation to the sample point to see if it's in the bounds.  Probably that is the better approach for all the transforms.  I.e. the transforms are all resampling transforms, taking world space points into the local space of the expressions.

Now that I think about it bounds should be isInBounds should be a virtual method that returns true by default but can be implemented for most expression types.  Then an _bounds method can calculate and cache the bounds of the subexpressions for efficient evaluation.

I'm thinking of the cases where there are nonlinear transforms and/or procedural expressions where isInBounds is easy to calculate but bounds is not.  

For some complex modeling tasks being able to trivially reject large portions of the grid will be great for efficiency!

**Comment by pieper** [comment]
Hey Stefan -

I put some notes in the line comments.  Let me know what you think.

I'm thinking something like

https://github.com/Kitware/VTK/blob/28aebaa3c3a722edd57275c35c4e7dbe1cf8cabc/Common/DataModel/vtkImplicitFunction.h

-Steve

**Comment by bebbi** [comment]
branch still exists for reference

**Pull request closed without merging**